x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java

```diff
@@ -363,6 +408,63 @@ public void clusterChanged(ClusterChangedEvent event) {
         }
     }
 
+    private void updateNumberOfAllocations(TrainedModelAssignmentMetadata assignments) {
+        List<TrainedModelAssignment> modelsToUpdate = assignments.modelAssignments()
+            .values()
+            .stream()
+            .filter(a -> hasStartingAssignments(a) == false)
+            .filter(a -> a.isRoutedToNode(nodeId))
+            .filter(a -> {
+                RoutingInfo routingInfo = a.getNodeRoutingTable().get(nodeId);
+                return routingInfo.getState() == RoutingState.STARTED
+                    && routingInfo.getCurrentAllocations() != routingInfo.getTargetAllocations();
+            })
+            .toList();
+
+        for (TrainedModelAssignment assignment : modelsToUpdate) {
+            TrainedModelDeploymentTask task = modelIdToTask.get(assignment.getModelId());
+            if (task == null) {
+                logger.debug(() -> format("[%s] task was removed whilst updating number of allocations", task.getModelId()));
+                continue;
+            }
+            RoutingInfo routingInfo = assignment.getNodeRoutingTable().get(nodeId);
+            deploymentManager.updateNumAllocations(
```

Comment:
This method doesn't necessarily call its listener.  That's probably OK, as the situation is a bit like `task == null` above where a removal is racing with the change.  I think it would be worth adding a comment here to say that, just in case in the future somebody makes the assumption that the listener will get called.  Better still, make `updateNumAllocations` pass `ResourceNotFoundException` to its listener for this case and explicitly ignore that in the error branch of the listener here.

Reply:
`DeploymentManager.updateNumAllocations` will indeed call the listener.onFailure if the process context is missing. It passes `listener::onFailure` in the error consumer of the `getProcessContext` method.